flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarProducer.java

```diff
@@ -0,0 +1,304 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.pulsar;
+
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.api.java.ClosureCleaner;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+import org.apache.flink.streaming.connectors.pulsar.partitioner.PulsarKeyExtractor;
+import org.apache.flink.util.SerializableObject;
+
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessageBuilder;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.ProducerConfiguration;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Flink Sink to produce data into a Pulsar topic.
+ */
+public class FlinkPulsarProducer<IN>
+		extends RichSinkFunction<IN>
+		implements CheckpointedFunction {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FlinkPulsarProducer.class);
+
+	/**
+	 * The pulsar service url.
+	 */
+	protected final String serviceUrl;
+
+	/**
+	 * User defined configuration for the producer.
+	 */
+	protected final ProducerConfiguration producerConfig;
+
+	/**
+	 * The name of the default topic this producer is writing data to.
+	 */
+	protected final String defaultTopicName;
+
+	/**
+	 * (Serializable) SerializationSchema for turning objects used with Flink into.
+	 * byte[] for Pulsar.
+	 */
+	protected final SerializationSchema<IN> schema;
+
+	/**
+	 * User-provided key extractor for assigning a key to a pulsar message.
+	 */
+	protected final PulsarKeyExtractor<IN> flinkPulsarKeyExtractor;
+
+	/**
+	 * Produce Mode.
+	 */
+	protected PulsarProduceMode produceMode = PulsarProduceMode.AT_LEAST_ONE;
+
+	/**
+	 * If true, the producer will wait until all outstanding records have been send to the broker.
+	 */
+	protected boolean flushOnCheckpoint;
+
+	// -------------------------------- Runtime fields ------------------------------------------
+
+	/** KafkaProducer instance. */
+	protected transient Producer producer;
+
+	/** The callback than handles error propagation or logging callbacks. */
+	protected transient Function<MessageId, MessageId> successCallback = msgId -> {
```

Comment:
This won't deserialise correctly, it will end up null. The solution is to initialise successCallback in the open() method instead, similar to how failureCallback is initialised.